saleor/account/models.py

```diff
@@ -189,6 +191,48 @@ def get_ajax_label(self):
         return self.email
 
 
+class Bot(models.Model):
+    name = models.CharField(max_length=128)
+    auth_token = models.CharField(default=generate_token, unique=True, max_length=30)
+    created = models.DateTimeField(auto_now_add=True)
+    is_active = models.BooleanField(default=True)
+    permissions = models.ManyToManyField(
+        Permission,
+        verbose_name=_("bot permissions"),
+        blank=True,
+        help_text=_("Specific permissions for this bot."),
+        related_name="bot_set",
+        related_query_name="bot",
+    )
+
+    class Meta:
+        permissions = (
+            ("manage_bots", pgettext_lazy("Permission description", "Manage bots")),
+        )
+
+    def _get_permissions(self):
+        """Return the permissions of the bot."""
+        if not self.is_active:
+            return set()
+        perm_cache_name = "_bot_perm_cache"
+        if not hasattr(self, perm_cache_name):
+            perms = self.permissions.all()
+            perms = perms.values_list("content_type__app_label", "codename").order_by()
+            setattr(self, perm_cache_name, {"%s.%s" % (ct, name) for ct, name in perms})
+        return getattr(self, perm_cache_name)
+
+    def has_perms(self, perm_list):
+        """Return True if the bot has each of the specified permissions."""
+        return all(self.has_perm(perm) for perm in perm_list)
```

Comment:
Very heavy... What an operation over a set?

```python
wanted_perms = set(perm_list)
actual_perms = self._get_permissions()

if not self.is_active:
    return False

return (wanted_perms & actual_perms) == wanted_perms

Reply:
It is not heavy. It operates on cached perms. I will change it anyway